src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.index.IndexConstants.OPTIMIZE_MODES
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, OptimizeActionEvent}
+import com.microsoft.hyperspace.util.{HyperspaceConf, PathUtils}
+
+/**
+ * Optimize Action provides an optimize support for indexes where small index files
+ * can be merged into larger ones for better index performance.
+ *
+ * Algorithm outline:
+ * 1. Collect all the currently valid index files.
+ * 2. Split files into small and large, based on a threshold.
+ * 3. Bucketwise combine smaller files into 1 file per bucket.
+ * 4. Update index snapshot to remove small files and keep large files + newly created files.
+ *
+ * NOTE: This is an index-only operation. It does not look at the current state of
+ * the data at all. If the data was changed after index creation, optimize will NOT
+ * include the data changes.
+ *
+ * Available modes:
+ * `Quick` mode: This mode allows for fast optimization. Files smaller than a
+ * predefined threshold "spark.hyperspace.index.optimize.fileSizeThreshold" will be picked
+ * for compaction.
+ *
+ * `Full` mode: This allows for slow but complete optimization. ALL files are
+ * eligible for compaction. Compaction threshold is inifinity.
+ *
+ * TODO: Optimize can be a no-op if there is only one small file per bucket.
+ *  https://github.com/microsoft/hyperspace/issues/204.
```

Comment:
Note to reviewers: this issue will handle the case where the data is already optimized (quick or full). For now, optimize will work if any "optimizable" files are found even if there is only one optimizable file per bucket.

Reply:
The size (small) shouldn't matter right? "Optimize can be a no-op if there is at most one file for every bucket"?

Reply:
Thanks @imback82, updated the comment.

it should be at most one "optimizable" file per bucket. This will be more general I think. it covers the "at most one file per bucket" scenario as well.

Reply:
OK. (I put at most because there can be buckets with zero files.)

Reply:
oh i understand now, i was missing the "at most" part. I will fix the comment to reflect this.